File: lib/src/domain/macd_calculator.dart
/// MACD Calculator — Pure domain logic.
///
/// Computes the MACD (Moving Average Convergence Divergence) indicator:
///   MACD Line  = EMA(12) - EMA(26)
///   Signal Line = EMA(9) of MACD Line
///   Histogram  = MACD Line - Signal Line
library;

import 'entities.dart';
import 'ema_calculator.dart';

class MacdResult {
  const MacdResult({
    required this.date,
    required this.macd,
    required this.signal,
    required this.histogram,
  });

  final DateTime date;

  /// MACD line value (null during warmup).
  final double? macd;

  /// Signal line value (null during warmup).
  final double? signal;

  /// Histogram value (null during warmup).
  final double? histogram;
}

class MacdCalculator {
  const MacdCalculator({this.fastPeriod = 12, this.slowPeriod = 26, this.signalPeriod = 9});

  final int fastPeriod;
  final int slowPeriod;
  final int signalPeriod;

  final _ema = const EmaCalculator();

  /// Compute the full MACD + Signal + Histogram series.
  List<MacdResult> computeSeries(List<DailyCandle> candles) {
    final fastSeries = _ema.computeSeries(candles, period: fastPeriod);
    final slowSeries = _ema.computeSeries(candles, period: slowPeriod);

    // Build the MACD line where both fast and slow are available
    final macdValues = <(DateTime, double?)>[];
    for (int i = 0; i < candles.length; i++) {
      final f = fastSeries[i].$2;
      final s = slowSeries[i].$2;
      macdValues.add((candles[i].date, (f != null && s != null) ? f - s : null));
    }

    // Extract non-null MACD values to compute Signal EMA via virtual candles
    final macdNonNull = <DailyCandle>[];
    final macdNullCount = macdValues.indexWhere((v) => v.$2 != null);
    if (macdNullCount < 0) {
      return candles
          .map((c) => MacdResult(date: c.date, macd: null, signal: null, histogram: null))
          .toList();
    }
    for (int i = macdNullCount; i < macdValues.length; i++) {
      macdNonNull.add(
        DailyCandle(
          date: macdValues[i].$1,
          open: macdValues[i].$2!,
          high: macdValues[i].$2!,
          low: macdValues[i].$2!,
          close: macdValues[i].$2!,
          volume: 0,
        ),
      );
    }

    final signalSeries = _ema.computeSeries(macdNonNull, period: signalPeriod);

    final result = <MacdResult>[];
    int signalIdx = 0;
    for (int i = 0; i < candles.length; i++) {
      final macdVal = macdValues[i].$2;
      if (macdVal == null || i < macdNullCount) {
        result.add(MacdResult(date: candles[i].date, macd: null, signal: null, histogram: null));
        continue;
      }
      final sig = signalSeries[signalIdx].$2;
      result.add(MacdResult(
        date: candles[i].date,
        macd: macdVal,
        signal: sig,
        histogram: (sig != null) ? macdVal - sig : null,
      ));
      signalIdx++;
    }
    return result;
  }

  /// Convenience: return the latest complete MACD result (or null).
  MacdResult? compute(List<DailyCandle> candles) {
    final series = computeSeries(candles);
    for (int i = series.length - 1; i >= 0; i--) {
      if (series[i].histogram != null) return series[i];
    }
    return null;
  }
}

File: src/domain/macd-calculator.ts
/**
 * MACD Calculator — Pure domain logic.
 * Ported from Dart: lib/src/domain/macd_calculator.dart
 *
 * MACD Line = EMA(fast) - EMA(slow)
 * Signal Line = EMA(signalPeriod) of MACD Line
 * Histogram = MACD Line - Signal Line
 */
import type { DailyCandle } from "../types/domain";
import { computeEmaSeries } from "./ema-calculator";
import { DEFAULTS } from "./technical-defaults";

export interface MacdPoint {
  readonly date: string;
  readonly macd: number | null;
  readonly signal: number | null;
  readonly histogram: number | null;
}

/**
 * Compute the full MACD + Signal + Histogram series.
 */
export function computeMacdSeries(
  candles: readonly DailyCandle[],
  fastPeriod: number = DEFAULTS.macdFastPeriod,
  slowPeriod: number = DEFAULTS.macdSlowPeriod,
  signalPeriod: number = DEFAULTS.macdSignalPeriod,
): MacdPoint[] {
  const fastSeries = computeEmaSeries(candles, fastPeriod);
  const slowSeries = computeEmaSeries(candles, slowPeriod);

  // Build MACD line where both fast and slow are available
  const macdValues: { date: string; value: number | null }[] = [];
  for (let i = 0; i < candles.length; i++) {
    const candle = candles[i];
    const fast = fastSeries[i];
    const slow = slowSeries[i];
    if (!candle || !fast || !slow) continue;
    const value = fast.value !== null && slow.value !== null ? fast.value - slow.value : null;
    macdValues.push({ date: candle.date, value });
  }

  // Extract non-null MACD values to compute Signal EMA
  const macdNullCount = macdValues.findIndex((v) => v.value !== null);
  if (macdNullCount < 0) {
    return candles.map((c) => ({
      date: c.date,
      macd: null,
      signal: null,
      histogram: null,
    }));
  }

  // Create virtual candles from MACD values for signal EMA computation
  const virtualCandles: DailyCandle[] = [];
  for (let i = macdNullCount; i < macdValues.length; i++) {
    const mv = macdValues[i];
    if (mv?.value !== null && mv?.value !== undefined) {
      virtualCandles.push({
        date: mv.date,
        open: mv.value,
        high: mv.value,
        low: mv.value,
        close: mv.value,
        volume: 0,
      });
    }
  }

  const signalSeries = computeEmaSeries(virtualCandles, signalPeriod);

  const result: MacdPoint[] = [];
  let signalIdx = 0;
  for (let i = 0; i < candles.length; i++) {
    const mv = macdValues[i];
    if (mv?.value == null || i < macdNullCount) {
      const candle = candles[i];
      if (candle) {
        result.push({ date: candle.date, macd: null, signal: null, histogram: null });
      }
      continue;
    }
    const sig = signalSeries[signalIdx];
    const sigVal = sig?.value ?? null;
    result.push({
      date: mv.date,
      macd: mv.value,
      signal: sigVal,
      histogram: sigVal !== null ? mv.value - sigVal : null,
    });
    signalIdx++;
  }

  return result;
}
